src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -141,12 +141,11 @@ protected static JsonConverter<TKey> GetKeyConverter(Type keyType, JsonSerialize
                 bool preserveReferences = options.ReferenceHandler != null;
                 if (preserveReferences && state.Current.ObjectState < StackFrameObjectState.PropertyValue)
                 {
-                    if (JsonSerializer.ResolveMetadataForJsonObject(ref reader, ref state, options, out string? referenceId))
+                    if (JsonSerializer.ResolveMetadataForJsonObject<TCollection>(ref reader, ref state, options))
                     {
                         if (state.Current.ObjectState == StackFrameObjectState.ReadRefEndObject)
                         {
-                            value = JsonSerializer.TryCastPreservedValue<TCollection>(
-                                state.Current.ReturnValue!, referenceId!);
+                            value = (TCollection)state.Current.ReturnValue!;
```

Comment:
nit: add comment saying that we validated that this won't throw `InvalidCastOperation` upstream. Here and in `IEnumerableDefaultConverter` below.